File: phys/fluid/HexGrid.java
package phys.fluid;

import util.Flags;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.Polygon;
import java.awt.image.BufferedImage;

/**
 * The HexGrid class provides a grid of hexagons that can be drawn
 * line-by-line, varying the color of each cell, for use in
 * visualizing the dynamics of hexagonal cellular automata.
 *
 * @see HexGridHorizontal
 * @see HexGridVertical
 */
abstract class HexGrid {

  static final boolean DRAW_BORDER = new Flags(HexGrid.class).get("drawBorder", true);

  BufferedImage img;
  Graphics g;
  int cols, rows, scale, count, xStrideScaled, yStrideScaled;
  int lineWidthPix;
  Polygon hex;
  final int []
    xc = new int[6],
    yc = new int[6];

  HexGrid(int cols, int rows, int scale,
          int [] XC, int [] YC,
          int xStride, int yStride) {
    System.out.printf("cols: %d, rows: %d, scale: %d\n", cols, rows, scale);
    this.cols = cols;
    this.rows = rows;
    this.scale = scale;
    for (int i = 0; i < 6; i++) {
      xc[i] = XC[i] * scale;
      yc[i] = YC[i] * scale;
    }
    hex = new Polygon(xc, yc, 6);
    xStrideScaled = xStride * scale;
    yStrideScaled = yStride * scale;
    lineWidthPix = cols * xStrideScaled;
    System.out.printf("xStrideScaled: %d, yStrideScaled: %d\n",
                      xStrideScaled, yStrideScaled);
    img = new BufferedImage(getWidth(), getHeight(), BufferedImage.TYPE_INT_RGB);
    g = img.getGraphics();
    count = 0;
  }

  abstract int getWidth();
  abstract int getHeight();

  void reset() {
    hex = new Polygon(xc, yc, 6);
    count = 0;
  }

  void next(Color c) {
    g.setColor(c);
    g.fillPolygon(hex);
    if (DRAW_BORDER) {
      g.setColor(Color.BLACK);
      g.drawPolygon(hex);
    }
    count++;
  }

  void line() {
    count = 0;
  }

  private void drawTest() {
    for (int i = 0, I = cols; i < I; i++) {
      for (int j = 0, J = rows; j < J; j++) {
        next(new Color(1f,
                       0.5f + (0.5f * i / (float)I),
                       0.5f + (0.5f * j / (float)J)));
      }
      line();
    }
  }
}

File: phys/fluid/HexGrid.dart
#library('phys/fluid/HexGrid');
//#import('../../util/Flags.dart');

/**
 * Hexagonal grid and image.
 *
 * @author Pablo Mayrgundter <pablo.mayrgundter@gmail.com>
 */
class HexGrid {

  final bool DRAW_BORDER = false;
  final double TAU = Math.PI * 2;

  /**
   * Coordinates of a polygon.
   */
  /*
  xc = new List<int>();
  yc = new List<int>();
  xc[0] = {1,3,4,3,1,0},
  yc = {0,0,2,4,4,2};
  */

  var g;
  int cols, rows, scale, count, xStride, yStride, yOffset;
  //Polygon hex;
  int x, y;
  int radius;

  HexGrid(int columns, int rows, int scale, var graphics) {
    this.cols = columns;
    this.rows = rows;
    this.scale = scale;
    this.g = graphics;
    /*
    for (int i = 0; i < 6; i++) {
      xc[i] = xc[i] * scale;
      yc[i] = yc[i] * scale;
    }
    */
    x = 0;
    y = 0;
    // hex = new Polygon(xc, yc, 6);
    xStride = 3 * scale;
    yStride = 2 * scale;
    yOffset = yStride;
    count = 0;
    radius = scale;
  }

  void drawTest() {
    var color = 'black';
    for (int i = 0, I = cols; i < I; i++) {
      for (int j = 0, J = rows; j < J; j++) {
        //next(new Color(0,
        // 0.5f + (0.5f * i / (float)I),
        // 0.5f + (0.5f * j / (float)J)));
        next(color);
      }
      line();
    }
  }

  void reset() {
    //hex = new Polygon(xc, yc, 6);
    x = 0;
    y = 0;
    yOffset = yStride;
    count = 0;
  }

  void next(String color) {
    g.beginPath();
    g.fillStyle = color;
    //g.fillPolygon(hex);
    g.arc(x, y, radius, 0, TAU, false);
    if (DRAW_BORDER) {
      g.fillStyle = 'blue';
      //g.drawPolygon(hex);
      var radius = 3;
      g.arc(x, y, radius, 0, TAU, false);
    }
    g.fill();
    g.closePath();
    g.stroke();
    x += xStride;
    y += yOffset;
    yOffset *= -1;
    count++;
  }

  void line() {
    // if yOffset is now negative, it was positive for last
    // translate, so only add half a yStride.  Otherwise, it was
    // negative for last translate, so add 1.5 strides.
    x += -1 * (count * xStride);
    y += yOffset < 0 ? yStride : 2 * yStride;
    yOffset = yOffset < 0 ? -yOffset : yOffset;
    count = 0;
  }
}
